src/compilation/build.py

```diff
@@ -0,0 +1,418 @@
+import os
+
+class CompilationArch(Enum):
+    ARM = "arm"
+    ARM64 = "aarch64"
+    POWERPC = "powerpc"
+    X64 = "x86_64"
+    MIPS = "mips"
+    MIPSEL = "mipsel"
+
+class BuildVariant(Enum):
+    SLIM = "slim"
+    FULL = "full"
+
+ARCH_HOSTS = {
+    CompilationArch.ARM: "arm-linux-musleabi",
+    CompilationArch.ARM64: "aarch64-linux-musl",
+    CompilationArch.POWERPC: "powerpc-linux-musl",
+    CompilationArch.MIPS: "mips-linux-musl",
+    CompilationArch.MIPSEL: "mipsel-linux-musl",
+    CompilationArch.X64: "x86_64-linux-musl",
+}
+
+@dataclass
+class CompilationEnv:
+    arch: CompilationArch
+    python: str
+
+    src_dir: Path
+    build_dir: Path
+
+    HOST: str = field(init=False)
+    CC: str = field(init=False)
+    CXX: str = field(init=False)
+    AR: str = field(init=False)
+    CFLAGS: List[str] = field(default_factory=list)
+    CXXFLAGS: List[str] = field(default_factory=list)
+    LDFLAGS: List[str] = field(default_factory=list)
+
+    def __post_init__(self):
+        if self.arch not in ARCH_HOSTS.keys():
+            raise RuntimeError(f"Architecture {self.arch} isn't supported")
+
+        self.HOST = ARCH_HOSTS[self.arch]
+
+        self.CC = str(f"{self.HOST}-gcc")
+        self.CXX = str(f"{self.HOST}-g++")
+        self.AR = str(f"{self.HOST}-ar")
+
+    @staticmethod
+    def format_env(env_words: List[str]) -> str:
+        return " ".join(shlex.quote(word) for word in env_words)
+
+    @property
+    def script_dir(self):
+        return self.src_dir / "compilation/"
+    
+    @property
+    def submodules_dir(self):
+        return self.src_dir / "submodule_packages/"
+    
+    @property
+    def packages_dir(self):
+        return self.build_dir / "packages/"
+
+    def as_env(self) -> Dict[str, str]:
+        return {
+            "HOST":     self.HOST,
+            "CC":       self.CC,
+            "CXX":      self.CXX,
+            "AR":       self.AR,
+            "CFLAGS":   self.format_env(self.CFLAGS),
+            "CXXFLAGS": self.format_env(self.CXXFLAGS),
+            "LDFLAGS":  self.format_env(self.LDFLAGS),
+        }
+
+    def add_compile_flag(self, *flags: str):
+        for flag in flags:
+            self.CFLAGS.append(flag)
+            self.CXXFLAGS.append(flag)
+
+    def add_flags(self, library: "LibraryMetadata"):
+        for lib_path in [library.get_install_path(self.arch) / "lib"] + library.additional_lib_paths:
+            self.LDFLAGS.append(f"-L{lib_path}")
+
+        for include_path in [library.get_install_path(self.arch) / "include"] + library.additional_include_paths:
+            self.add_compile_flag(f"-I{include_path}")
+
+@dataclass
+class LibraryMetadata:
+    source_path: Path
+    commands: List[str]
+
+    # If relative then it's relative to the install path.
+    install_path: Path = Path("output")
+
+    env_setup: Callable[[CompilationEnv, Dict[str, str]], None] = lambda _, __: None
+
+    # If relative then it's relative to the install path.
```

Comment:
Same here, code implies its always to the install path
```suggestion
    # Relative to install path.
```

Reply:
https://github.com/guyush1/gdb-static/pull/88#discussion_r2649748601